packages/flutter_tools/bin/xcode_debug.js

```diff
@@ -0,0 +1,519 @@
+/**
+ * Curated results from Xcode's debug function. Mirrors parts of
+ * `scheme action result` from Xcode's Script Editor dictionary.
+ */
+class DebugResult {
+  /**
+   *
+   * @param {!Object} result
+   */
+  constructor(result) {
+    this.completed = result.completed();
+    this.status = result.status();
+    this.errorMessage = result.errorMessage();
+  }
+}
+
+/**
+ * Get the Xcode application from the given path. Since macs can have multiple
+ * Xcode version, we use the path to target the specific Xcode application.
+ * If the Xcode app is not running, return null with an error.
+ *
+ * @param {!CommandArguments} args
+ * @returns {!FunctionResult} Return either an `Application` (Mac Scripting class)
+ *     or null as the `result`.
+ */
+function getXcode(args) {
+  try {
+    const xcode = Application(args.xcodePath);
+    const isXcodeRunning = xcode.running();
+
+    if (isXcodeRunning === false) {
+      return new FunctionResult(null, 'Xcode is not running');
+    }
+
+    return new FunctionResult(xcode);
+  } catch (e) {
+    return new FunctionResult(null, `Failed to get Xcode application: ${e}`);
+  }
+}
+
+/**
+ * After setting the active run destination to the targeted device, uses Xcode
+ * debug function from Mac Scripting for Xcode to install the app on the device
+ * and start a debugging session using the 'run' or 'run without building' scheme
+ * action (depending on `args.skipBuilding`). Waits for the debugging session
+ * to start running.
+ *
+ * @param {!Application} xcode An `Application` (Mac Scripting class) for Xcode.
+ * @param {!CommandArguments} args
+ * @returns {!FunctionResult} Return either a `DebugResult` or null as the `result`.
+ */
+function debugApp(xcode, args) {
+  const workspaceResult = waitForWorkspaceToLoad(xcode, args);
+  if (workspaceResult.error != null) {
+    return new FunctionResult(null, workspaceResult.error);
+  }
+	const targetWorkspace = workspaceResult.result;
+
+  const destinationResult = getTargetDestination(
+    targetWorkspace,
+    args.targetDestinationId,
+    args.verbose,
+  );
+  if (destinationResult.error != null) {
+    return new FunctionResult(null, destinationResult.error)
+  }
+
+  try {
+    // Documentation from the Xcode Script Editor dictionary indicates that the
+    // `debug` function has a parameter called `runDestinationSpecifier` which
+    // is used to specify which device to debug the app on. It also states that
+    // it should be the same as the xcodebuild -destination specifier. It also
+    // states that if not specified, the `activeRunDestination` is used instead.
+    //
+    // Experimentation has shown that the `runDestinationSpecifier` does not work.
+    // It will always use the `activeRunDestination`. To mitigate this, we set
+    // the `activeRunDestination` to the targeted device prior to starting the debug.
+    targetWorkspace.activeRunDestination = destinationResult.result;
+
+    const actionResult = targetWorkspace.debug({
```

Comment:
is this `debug` API provided by apple? can i see a link?

Reply:
Yes, it's an API provided by Apple. There's not any published documentation that I know of. But you can open the app `Script Editor` on your mac and then in the top menu bar, select File > Open Dictionary, click on Xcode, click Choose button and it'll open a window with the documentation.